### PR TITLE
AuRa Producing refactoring and fixes

### PR DESCRIPTION
**Pull request by LukaszRozmej**

### CODE REVIEW

File: src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs [hunk 1]
```diff
@@ -54,9 +54,23 @@ protected Transaction GenerateTransaction(byte[] transactionData, long gasLimit
         
         public void InvokeTransaction(BlockHeader header, ITransactionProcessor transactionProcessor, Transaction transaction, CallOutputTracer tracer)
         {
-            if (transaction != null && !TryInvokeTransaction(header, transactionProcessor, transaction, tracer))
+            if (transaction != null)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs]: why would the transaction ever be null in such a call?

File: src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs [hunk 2]
```diff
@@ -54,9 +54,23 @@ protected Transaction GenerateTransaction(byte[] transactionData, long gasLimit
...
             {
-                throw new AuRaException($"System call returned error '{tracer.Error}' at block {header.Number}.");
+                bool failure;
+                
+                try
+                {
+                    failure = !TryInvokeTransaction(header, transactionProcessor, transaction, tracer);
+                }
+                catch (Exception e)
+                {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs]: usually ine does not expect the Try method to throw an exception

File: src/Nethermind/Nethermind.AuRa/Validators/MultiValidator.cs [hunk 3]
```diff
@@ -100,7 +102,7 @@ public void PreProcess(Block block, ProcessingOptions options = ProcessingOption
             {
                 if (TryGetLastValidator(previousBlockNumber, out var validatorInfo))
                 {
-                    if (validatorInfo.Value.ValidatorType.CanChangeImmediately())
+                    if (validatorInfo.Value.ValidatorType.CanChangeImmediately() || _blockFinalizationManager.LastFinalizedBlockLevel >= validatorInfo.Key)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Validators/MultiValidator.cs]: can we add some comment / explainer here

File: src/Nethermind/Nethermind.AuRa/Validators/MultiValidator.cs [hunk 4]
```diff
@@ -178,9 +180,13 @@ private void SetCurrentValidator(KeyValuePair<long, AuRaParameters.Validator> va
         
         private void SetCurrentValidator(long finalizedAtBlockNumber, AuRaParameters.Validator validator)
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Validators/MultiValidator.cs]: shall we rename validator argument to newValidatorPrototype

File: src/Nethermind/Nethermind.Blockchain/BaseLoopBlockProducer.cs [hunk 5]
```diff
@@ -68,15 +68,17 @@ public override async Task StopAsync()
             await (_producerTask ?? Task.CompletedTask);
         }
         
-        private async ValueTask ProducerLoop()
+        protected virtual async ValueTask ProducerLoop()
         {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/BaseLoopBlockProducer.cs]: CancellationToken

File: src/Nethermind/Nethermind.Blockchain/DevBackgroundBlockProducer.cs [hunk 6]
```diff
@@ -51,8 +51,9 @@ public class DevBackgroundBlockProducer : BaseLoopBlockProducer
 
         protected override UInt256 CalculateDifficulty(BlockHeader parent, UInt256 timestamp) => 1;
 
-        protected override async ValueTask BetweenBlocks()
+        protected override async ValueTask ProducerLoopStep()
         {
+            await base.ProducerLoopStep();
             await Task.Delay(DelayBetweenBlocks);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/DevBackgroundBlockProducer.cs]: pass cancellation token to delay

File: src/Nethermind/Nethermind.Blockchain/DevBackgroundBlockProducer.cs [hunk 7]
```diff
@@ -51,8 +51,9 @@ public class DevBackgroundBlockProducer : BaseLoopBlockProducer
 
         protected override UInt256 CalculateDifficulty(BlockHeader parent, UInt256 timestamp) => 1;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/DevBackgroundBlockProducer.cs]: is it really always 1?

File: src/Nethermind/Nethermind.Core/BlockHeader.cs [hunk 8]
```diff
@@ -60,7 +60,7 @@ public BlockHeader(Keccak parentHash, Keccak ommersHash, Address beneficiary, UI
         public long GasUsed { get; set; }
         public long GasLimit { get; set; }
         public UInt256 Timestamp { get; set; }
-        public DateTime TimestampDate => DateTimeOffset.FromUnixTimeSeconds((long) Timestamp).DateTime;
+        public DateTime TimestampDate => DateTimeOffset.FromUnixTimeSeconds((long) Timestamp).LocalDateTime;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Core/BlockHeader.cs]: Why local and not UTC?

File: src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs [hunk 9]
```diff
@@ -116,13 +117,16 @@ public override void PreProcess(Block block, ProcessingOptions options = Process
         {
             var isProducingBlock = options.IsProducingBlock();
             var isProcessingBlock = !isProducingBlock;
+            var initBlock = InitBlockNumber == block.Number;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs]: isInitBlock

File: src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs [hunk 10]
```diff
@@ -116,13 +117,16 @@ public override void PreProcess(Block block, ProcessingOptions options = Process
...
+            var loadValidators = _validators == null || isProducingBlock;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs]: shouldLoadValidators

File: src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs [hunk 11]
```diff
@@ -116,13 +117,16 @@ public override void PreProcess(Block block, ProcessingOptions options = Process
...
             
-            if (_validators == null || isProducingBlock)
+            if (loadValidators)
             {
                 Validators = LoadValidatorsFromContract(block.Header);
+                if(_logger.IsInfo && !_isProducing && isProcessingBlock) _logger.Info($"{(initBlock ? "Initial" : "Current")} contract validators ({Validators.Length}): [{string.Join<Address>(", ", Validators)}].");
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs]: do not mix if(logger.isLevel) with business logic -may easy lead to errors

Comment (tkstanczak) [reply src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs]: why do we have _isProducing and isProducingBlock? can we make them more easily distinguished?

Comment (tkstanczak) [reply src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs]: now we have a strange looking !isProducing && isProcessing and the first impression is that one of them is always true. not abvious that one is a condig for isMining and other for current run

File: src/Nethermind/Nethermind.Blockchain/DevBackgroundBlockProducer.cs (hunk 7 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Blockchain/DevBackgroundBlockProducer.cs]: its dev-test only

File: src/Nethermind/Nethermind.Core/BlockHeader.cs (hunk 8 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/BlockHeader.cs]: This is later used only for logging purposes, all other logs are in local time. We can have it in local time here, or convert to local time on logging. But like I said currently logging is the only use case.

File: src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs]: Not really, unless dev error. Will remove.

File: src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AuRa/Contracts/SystemContract.cs]: True, rewriting...

File: src/Nethermind/Nethermind.AuRa/Validators/MultiValidator.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AuRa/Validators/MultiValidator.cs]: Will name it in variable

File: src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs (hunk 11 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs]: changing structure and doing a little renaming.

**Pull request merged**